symengine and python-symengine packages

isuruf opened the pull request:
cc @certik

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I wanted to let you know that I linted all conda-recipes in your PR (```recipes/python-symengine```, ```recipes/symengine```) and found some lint.

Here's what I've got...


For **recipes/python-symengine**:

 * The top level meta keys are in an unexpected order. Expecting ['package', 'source', 'build', 'requirements', 'test', 'about', 'extra'].

For **recipes/symengine**:

 * The top level meta keys are in an unexpected order. Expecting ['package', 'source', 'build', 'requirements', 'test', 'about', 'extra'].

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipes/python-symengine```, ```recipes/symengine```) and found it was in an excellent condition.

isuruf commented:
@jakirkham, can you review?

ocefpaf commented on recipes/python-symengine/meta.yaml:
```diff
@@ -0,0 +1,37 @@
+{% set version = "0.1.0.git" %}
+
```
You can add the commit as a jinja variable here:

`{% set commit = "4361e3782ce2b61c9962f3f9a1e766e481a00bf9" %}`

ocefpaf commented on recipes/python-symengine/meta.yaml:
```diff
@@ -0,0 +1,37 @@
+
+source:
+  git_url: https://github.com/symengine/symengine.py
+  git_tag: 4361e3782ce2b61c9962f3f9a1e766e481a00bf9
```
and then you can change this to:

```
source: yaml
    fn: {{ commit }}.tar.gz
    url: https://github.com/symengine/symengine.py/archive/{{ commit }}.tar.gz
    sha256: a28803ffbb15d4b252dad18d252d79419d1d561c1ead9ad1671821527c1f94a2
```

ocefpaf commented on recipes/symengine/meta.yaml:
```diff
@@ -0,0 +1,40 @@
+
+source:
+  git_url: https://github.com/symengine/symengine
+  git_tag: 3f58acc1246661323b1daf2adc3f9fa6a8762df4
```
Pleas do the same as above. (Also, can you ping the devs and request an official release?)

isuruf replied:
Fixed. We'll make an official release soon.

jakirkham commented on recipes/python-symengine/meta.yaml:
```diff
@@ -0,0 +1,37 @@
+  number: 0
+
+requirements:
+  build:
```
Please add `toolchain` here.

jakirkham commented on recipes/python-symengine/build.sh:
```diff
@@ -0,0 +1,18 @@
+
+    # This enables Xcode support for C++11, that is needed for SymEngine.
+    export MACOSX_DEPLOYMENT_TARGET=10.9
+fi
```
Please drop all of this and replace with `source activate "${CONDA_DEFAULT_ENV}"`.

Are you using C++11 on both platforms? If so, add this to the `$CXXFLAGS` independent of platform and make sure to keep the existing flags included in `$CXXFLAGS`.

isuruf replied:
Is, `export MACOSX_DEPLOYMENT_TARGET=10.9` added by default when you do `source activate "${CONDA_DEFAULT_ENV}"`?

jakirkham replied:
No, it sets it to 10.7, which is what we should be doing. It will still be compatible with 10.9. I use 10.9 regularly with binaries built here.

jakirkham commented on recipes/symengine/meta.yaml:
```diff
@@ -0,0 +1,40 @@
+  number: 0
+
+requirements:
+  build:
```
Please add `toolchain`.

jakirkham commented on recipes/symengine/build.sh:
```diff
@@ -0,0 +1,34 @@
+
+    # This enables Xcode support for C++11, that is needed for SymEngine.
+    export MACOSX_DEPLOYMENT_TARGET=10.9
+fi
```
Please drop all of this and replace with `source activate "${CONDA_DEFAULT_ENV}"`.

Are you using C++11 on both platforms? If so, add this to the `$CXXFLAGS` independent of platform and make sure to keep the existing flags included in `$CXXFLAGS`.

jakirkham commented on recipes/python-symengine/meta.yaml:
```diff
@@ -0,0 +1,37 @@
+  recipe-maintainers:
+    - isuruf
+    - certik
+
```
nit: please keep one terminal newline.

jakirkham commented on recipes/symengine/meta.yaml:
```diff
@@ -0,0 +1,40 @@
+  recipe-maintainers:
+    - isuruf
+    - certik
+
```
nit: please keep one terminal newline.

jakirkham commented:
Thanks @isuruf. I added a few notes as we have made some changes to how we do builds on Mac. In particular, we have a [`toolchain`]( https://github.com/conda-forge/toolchain-feedstock ) package, which helps line up the correct compiler on Mac. We may explore uses for other OSes, as well.

isuruf commented:
@jakirkham, I don't think travis-ci has Mac OSX 10.7 sdk. I get this warning when I remove `export MACOSX_DEPLOYMENT_TARGET=10.9`

```
Make Warning at /Users/travis/miniconda/envs/_build/share/cmake-3.5/Modules/Platform/Darwin-Initialize.cmake:97 (message):
  CMAKE_OSX_DEPLOYMENT_TARGET is '10.7' but the matching SDK does not exist
  at:
   "/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.7.sdk"
  Instead using SDK:
   "/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.9.sdk".
Call Stack (most recent call first):
  /Users/travis/miniconda/envs/_build/share/cmake-3.5/Modules/CMakeSystemSpecificInitialize.cmake:18 (include)
  CMakeLists.txt:2 (project)
```

isuruf commented:
I've removed `export MACOSX_DEPLOYMENT_TARGET=10.9` calls and you can see the warning yourself in travis-ci log.

certik commented:
@isuruf how can I test this PR before it gets merged? I can test if it builds.

isuruf commented:
@certik,
Add `conda-forge` channel
Install `conda-build-all`
Run `conda-build-all ./recipes --matrix-condition "numpy >=1.10" "python >=2.7,<3|>=3.4"`

certik commented:
I assume I check out this PR first locally and run this in the `staged-recipes` directory?

isuruf commented:
Yes

isuruf commented:
@jakirkham, this might be a silly question, have you used binaries from conda-forge in OS X 10.7? Do they work?

isuruf commented:
ping

jakirkham commented:
It's a good point you raise, @isuruf. Unfortunately I don't have 10.7 so cannot test this, but I suspect that your findings suggest that our binaries will not work on 10.7. That being said, they should be buildable on that platform and this is something I believe Continuum wants (though we will need to discuss). I have opened this issue ( https://github.com/conda-forge/toolchain-feedstock/issues/7 ) to evaluate different approaches that we could pursue.

jakirkham commented on recipes/python-symengine/build.sh:
```diff
@@ -0,0 +1,5 @@
+
+source activate "${CONDA_DEFAULT_ENV}"
+
+python setup.py install --symengine-dir=$PREFIX
```
Are there some tests that we can run?

isuruf replied:
Yes, I'll add them

jakirkham replied:
Thanks. Also, please add these arguments to your install command `--single-version-externally-managed --record record.txt`.

jakirkham commented on recipes/python-symengine/meta.yaml:
```diff
@@ -0,0 +1,39 @@
+{% set version = "0.1.0.git" %}
```
Could we please have [PEP440]( https://www.python.org/dev/peps/pep-0440/ ) compliant version number? Perhaps changing `0.1.0.git` to `0.1.0.dev0`.

jakirkham commented on recipes/symengine/meta.yaml:
```diff
@@ -0,0 +1,42 @@
+{% set version = "0.1.0.git" %}
```
Could we please have [PEP440]( https://www.python.org/dev/peps/pep-0440/ ) compliant version number? Perhaps changing `0.1.0.git` to `0.1.0.dev0`.

jakirkham commented:
We should start thinking about pinning `gmp`, `mpfr`, and `mpc`. There are several things that are going to dependent and linked to these; so, we should decide how we want to do this. Normally we do something like `<MAJOR>.<MINOR>.*`, but that is not always appropriate. Let's discuss how we want to do this.

ocefpaf commented on recipes/python-symengine/meta.yaml:
```diff
@@ -0,0 +1,39 @@
+    - cmake
+    - symengine
+    - python
+    - cython 0.23*
```
@isuruf Do you need to pin cython or `cython >= 0.23` would suffice here?

isuruf replied:
Cython 0.24 has some bugs which makes the Cython compiler crash. So need to use 0.23

jakirkham replied:
What kind of bugs?

isuruf replied:
One is https://mail.python.org/pipermail/cython-devel/2016-April/004771.html
There are some more errors when I remove the offending lines from the cython file.

jakirkham replied:
Thanks for sharing. This seems like a sensible solution to me.

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I wanted to let you know that I linted all conda-recipes in your PR (```recipes/python-symengine```, ```recipes/symengine```) and found some lint.

Here's what I've got...


For **recipes/symengine**:

 * Selectors are suggested to take a ``<two spaces>#<one space>[<expression>]`` form.

isuruf commented on recipes/python-symengine/meta.yaml:
```diff
@@ -0,0 +1,43 @@
+    - cython 0.23*
+  run:
+    - symengine
+    - python
```
For windows build, should there be `vc 14` here?

jakirkham replied:
Normally, we handle Windows via [this]( https://github.com/conda-forge/staged-recipes/wiki/VC-features ).

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipes/python-symengine```, ```recipes/symengine```) and found it was in an excellent condition.

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I wanted to let you know that I linted all conda-recipes in your PR (```recipes/python-symengine```, ```recipes/symengine```) and found some lint.

Here's what I've got...


For **recipes/python-symengine**:

 * Selectors are suggested to take a ``<two spaces>#<one space>[<expression>]`` form.

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipes/python-symengine```, ```recipes/symengine```) and found it was in an excellent condition.

isuruf commented:
I've addressed all of the comments, except pinning versions of gmp, mpfr and mpc. I have no idea what would make sense here.

There seems to be a regression in `conda-build` or `conda-build-all` because appveyor fails to skip checking dependencies for  `win and not py35` even though `win and not py35` is skipped. `conda-build==1.20` seems to work (This is used in py34 and win64). `conda-build==1.20.3` doesn't work. It tries to find a `symengine` package for `python-symengine` when `win and not py35` is true, but obviously, `symengine` doesn't exist for `win and not py35`

ocefpaf commented:
> I've addressed all of the comments, except pinning versions of gmp, mpfr
and mpc. I have no idea what would make sense here.

Thanks!

> There seems to be a regression in conda-build or conda-build-all because
appveyor fails to skip checking dependencies for win and not py35 even
though win and not py35 is skipped. conda-build==1.20 seems to work (This
is used in py34 and win64). conda-build==1.20.3 doesn't work. It tries to
find a symengine package for python-symengine when win and not py35 is
true, but obviously, symengine doesn't exist for win and not py35

The issue is with conda-build-all that does not know how to use the latest
conda-build. But downgrading conda-build helps. Let me merge a PR fixing
that and we can restart the CIs here.

ocefpaf commented:
@isuruf I am on my phone, can you try closing/opening the pr to get get AppVeyor goin?

ocefpaf commented:
@isuruf that the CIs config is in a package we need a new release to propagate the changes. I cannot do it today, but I'll do it ASAP tomorrow.

ocefpaf commented:
@isuruf AppVeyor should be OK now and the failure in the `Environment: TARGET_ARCH=x86, CONDA_PY=35, PY_CONDITION=python >=3.5` should be "real" :smile: 

Let me know if there is anything else I can do to help you here.

jakirkham commented:
Maybe we could try nmake and see if the problem persist.

isuruf commented on recipes/symengine/bld.bat:
```diff
@@ -0,0 +1,19 @@
+cd build
+
+cmake ^
+    -G "%CMAKE_GENERATOR%" ^
```
Problem is here, earlier the quotes were needed, now it's not.

isuruf replied:
I guess this change would break quite a number of build

isuruf replied:
`set "CMAKE_GENERATOR=Visual Studio 14 2015"` was changed to
`set CMAKE_GENERATOR="Visual Studio 14 2015"`

Using `cmake -G "%CMAKE_GENERATOR%"` worked earlier, now it should be changed to `cmake -G %CMAKE_GENERATOR%`.

ocefpaf replied:
> Using `cmake -G "%CMAKE_GENERATOR%"` worked earlier, now it should be changed to `cmake -G %CMAKE_GENERATOR%`.

Sorry @isuruf I don't have enough experience with `cmake` nor Windows to help you there.

isuruf replied:
It's all about the quotes. Earlier `CMAKE_GENERATOR` was `Visual Studio 14 2015` (without quotes). So when using in cmake, quotes were added `"%CMAKE_GENERATOR%"`.

Now, `CMAKE_GENERATOR` is `"Visual Studio 14 2015"` (with quotes). So, previous recipes with `"%CMAKE_GENERATOR%"` would become `""Visual Studio 14 2015""` (Note the double double-quotes). CMake doesn't work with the double double-quotes.

I just wanted to confirm that this was an intentional change because all recipes adding single double quotes would now have double double-quotes and would break.

jakirkham replied:
@patricksnape, can you please enlighten us on the quote choices made with `%CMAKE_GENERATOR%`?

patricksnape replied:
I believe @msarahan updated the quoting of this? I remember discussing this before somewhere.

ocefpaf replied:
> I believe @msarahan updated the quoting of this? I remember discussing this before somewhere.

If this has something to do with `conda-build` (sorry for my Windows ignorance here :grimacing:) remember that we are downgrading `conda-build` to `1.20.1` everywhere but Win64_PY34, where we are using `1.20.0`.

Those downgrades will be removed as soon as we fix https://github.com/SciTools/conda-build-all/issues/41

isuruf replied:
Yes, you are right. This [change](https://github.com/conda/conda-build/commit/d2d5d8a62d719485717700637f0f8d2394bffa23) is not in `1.20.1`

isuruf commented:
Thanks for the help @ocefpaf.

isuruf commented:
Tests pass now

jakirkham commented on recipes/symengine/meta.yaml:
```diff
@@ -0,0 +1,52 @@
+    - mpfr          # [unix]
+    - mpc           # [unix]
+    - mpir          # [win]
+    - python        # [win]
```
Does `symengine` actually use `python`? If not, this can be dropped from `run`.

isuruf replied:
No, it doesn't.
Shall I replace with `vc 14`?

jakirkham replied:
Ah, sorry, this was suppose to be on the one below. No, it shouldn't be needed. Though if you would prefer to try `vc` go for it.

isuruf replied:
I'm a little confused. I don't want to use `vc 14` right now. What should I do in this case?

jakirkham replied:
Please remove `python`.

jakirkham replied:
Sorry, I gave you too many options. Just remove `python` from `run` as shown below.

jakirkham replied:
Done.

jakirkham commented on recipes/symengine/meta.yaml:
```diff
@@ -0,0 +1,52 @@
+    - mpir          # [win]
+    - python        # [win]
+
+test:
```
Could we please add a block like the one below to `test`?

```
  requires:
    - python 2.7.*  # [win and py27]
    - python 3.4.*  # [win and py34]
    - python 3.5.*  # [win and py35]
```

isuruf replied:
```
  requires:
    - python 3.5.*  # [win and py35]
```
or
```
  requires:
    - vc 14  # [win and py35]
```
?

jakirkham replied:
Right, sorry, forgot you are using only 1 Python. The first one looks correct.

jakirkham replied:
Yeah, this has no effect on an install of `symengine` in the wild. It is only helpful to make sure testing proceeds correctly.

jakirkham replied:
Sorry, I made this confusing too. Just add...

```
  requires:
    - python 3.5.*  # [win and py35]
```

jakirkham commented on recipes/python-symengine/bld.bat:
```diff
@@ -0,0 +1 @@
+python setup.py install --symengine-dir=%LIBRARY_PREFIX%
```
Please add these arguments to your install command `--single-version-externally-managed --record record.txt`.

jakirkham commented on recipes/python-symengine/meta.yaml:
```diff
@@ -0,0 +1,47 @@
+
+requirements:
+  build:
+    - toolchain
```
Please add `setuptools`.

jakirkham commented:
Added some tweaks for the Windows build of `symengine` and some tweaks for the python installs (given you are using `setuptools` if found). Otherwise this looks really good, @isuruf. With those changes made, I would be +1 to merging.

isuruf commented:
Tests pass now

isuruf commented:
ping

jakirkham commented:
Are you alright being added as a maintainer on these, @certik?

jakirkham commented:
LGTM.

Would like a confirmation from @certik about being a maintainer.

Do you want to figure out `mpc` first and try Windows here or would you rather try that in the feedstock? Either way you chose, I'm ok with.

certik commented:
@jakirkham yes, that is fine with me. Thanks @isuruf for the work and @jakirkham for the review.

ocefpaf commented on recipes/python-symengine/build.sh:
```diff
@@ -0,0 +1,5 @@
+#!/usr/bin/env bash
+
+source activate "${CONDA_DEFAULT_ENV}"
```
Question: did you try removing this?

jakirkham replied:
We can't remove this yet as we are using too old of a `conda-build`.

jakirkham replied:
This would require this PR ( https://github.com/conda/conda-build/pull/1002 ), which I don't believe shows up in a release yet.

msarahan replied:
That's confusing. I did merge it into the 1.21.x branch before tagging it. https://github.com/conda/conda-build/commit/63b4db5826bcc6bf77242fc4e53caa3afc27cf40 

I may have done it clumsily. I'm new to managing multiple release branches.

isuruf commented:
Thanks @jakirkham. I'll add in MPC for windows in the feedstock after this has been merged in

jakirkham commented:
Thanks @isuruf. 👍 on merging for me.

More questions @ocefpaf or can we take this to a feedstock?

ocefpaf commented:
> More questions @ocefpaf or can we take this to a feedstock?

LGTM. Thanks @isuruf!

jakirkham commented:
Taking it in. Thanks everyone.

jakirkham commented:
@certik, as this is your first time being added as a maintainer on a recipe I would like to welcome you to conda-forge and briefly explain how things work.

When feedstock generation of these recipes occur, it will note that your GitHub handle has been added to the maintainers list. The feedstock generation script will try to add you to maintenance teams that give you permissions over the two feedstocks (repos) with these recipes. However, as you are not a member of conda-forge yet, GitHub will send you an email inviting you to join conda-forge. Accepting this will add you to these teams and provide you commit permissions.

When making changes to a feedstock (repo), please go through the normal process of forking the repo, making changes in a branch on your fork, and PR-ing those changes back.

If you have any questions or concerns, please let us know.

Welcome to conda-forge. 😄

certik commented:
Thanks @jakirkham, I appreciate it. The process is clear.
